spark/BidResponseJob.scala

```diff
@@ -0,0 +1,191 @@
+    .getGroupAggregates(
+      momentsColumns= {
+        case i => Map(
+          "spend" -> i.advertiserSpent.getOrElse(0.0),
+          "clearingPrice" -> i.clearingPrice.getOrElse(0.0),
+          "expProfit" -> Seq(Some(0.0), i.actualDSPFee, i.actualPGMFee, i.actualSSPFee, i.thirdpartyMargin).flatMap(e => e).sum
+        )
+      },
+      sumColumns={ case i => if (i.bidderId.isDefined) Map("isRtb"->1) else Map() },
+      countColumns= {
+        case i => Map(
+          "pricingTypes" -> i.pricingType.filterNot(_.isEmpty).getOrElse("UNKNOWN-PRICINGTYPE"),
+          "deviceTypes" -> i.userAgent.map(LogParser.parseUa(_).getDeviceCategory().getName()).filterNot(dt => dt.isEmpty || dt.toLowerCase() == "unknown").getOrElse("UNKNOWN-DEVICETYPE"),
+          "oss" -> i.userAgent.map(LogParser.parseUa(_).getOperatingSystem().getFamilyName()).filterNot(os => os.isEmpty || os.toLowerCase() == "unknown").getOrElse("UNKNOWN-OS"),
+          "browsers" -> i.userAgent.map(LogParser.parseUa(_).getFamily().getName()).filterNot(b => b.isEmpty || b.toLowerCase() == "unknown").getOrElse("UNKNOWN-BROWSER"),
+          "sourcesImp" -> i.buyeruidSource.filterNot(_.isEmpty).getOrElse("UNKNOWN-UIDSOURCE")
+        )
+      }
+    )
+
+  private lazy val processingDate = jobDates.lastDay.toStrings().head
+
+  purchaseGroups
+    .map{ case (PrefixGroup(bidderId,prefixGroup), s) =>
+      (bidderId, List(BidderGroupInfo(
+        date=processingDate,
+        definition=BidderGroupDefinition(bidderId, prefixGroup, "NO-LABEL", "NO-DESCRIPTION"),
+        bids=s.meanStatistic("bidPrice"),
+        nonZeroBids=s.meanStatistic("nonZeroBidPrice"),
+        largeBids=s.meanStatistic("largeBidPrice"),
+        spend=s.meanStatistic("spend"),
+        clearingPrice=s.meanStatistic("clearingPrice"),
+        expectedProfit=s.meanStatistic("expProfit"),
+        clicks=s.proportionStatistic("clicks", "spend"),
+        conversions=s.proportionStatistic("conversions", "spend"),
+        rtbWins=s.proportionStatistic("isRtb", "spend"),
+        sources=(if (bidderId == "ALL") s.categoryStatistic("sourcesImp", "spend") else s.categoryStatistic("sourcesBids", "bidPrice")),
+        pricingTypes=s.categoryStatistic("pricingTypes", "spend"),
+        responses=s.proportionStatistic("nonZeroBidPrice", "bidPrice"),
+        deviceTypes=s.categoryStatistic("deviceTypes", "spend"),
+        oss=s.categoryStatistic("oss", "spend"),
+        browsers=s.categoryStatistic("browsers", "spend"),
+        idaasIndices=s.categoryStatistic("idaasIndex", "bidPrice")
+      ) )) }
+    .reduceByKey{case (bgInfo1, bgInfo2) => {
+      val bgInfos1 = bgInfo1.reduce((b1, b2) => b1.crossWith(b2))
+      val bgInfos2 = bgInfo2.reduce((b1, b2) => b1.crossWith(b2))
+      bgInfo1.map(bgi1 => bgi1.crossWith(bgInfos2)) ++ bgInfo2.map(bgi2 => bgi2.crossWith(bgInfos1))
+    }}
+    .flatMap(_._2)
+    .coalesce(1)
+    .writeTo(BidderGroupInfoSource())
+
+}
+
+object BidResponsesJob {
```

Comment:
The stuff in this is rather hard to understand. I would see it as extremely impressive if someone is able to on their first read through. It could be made easier by adding some comments in there, and maybe rearranging. 
Question is if you want only the really impressive people, and want to be able to distinguish them from the rest (keep it as it is), or if you only need to distinguish the "good enough" people from the rest (add some help in there).

Reply:
i think for the current candidate, my thinking is to check whether she is "really impressive"? -- @nielsboldt what is your opinion? 
I have not thought about others, but my basic idea is the example code should be adapt a bit to each candidate